File: src/lib/parseTable.ts
export type ColumnRole = 'date' | 'description' | 'amount' | 'ignore';
export type DelimiterOption = 'auto' | ',' | ';' | '\t';

function splitLine(line: string, delimiter: string): string[] {
  const out: string[] = [];
  let cur = '';
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (inQuotes) {
      if (c === '"') {
        if (line[i + 1] === '"') { cur += '"'; i++; } else inQuotes = false;
      } else {
        cur += c;
      }
    } else if (c === '"') {
      inQuotes = true;
    } else if (c === delimiter) {
      out.push(cur.trim());
      cur = '';
    } else {
      cur += c;
    }
  }
  out.push(cur.trim());
  return out;
}

function mostCommonCount(counts: number[]): number {
  const tally = new Map<number, number>();
  for (const n of counts) tally.set(n, (tally.get(n) ?? 0) + 1);
  return Math.max(...tally.values());
}

function detectDelimiter(lines: string[]): string {
  let best = ',';
  let bestScore = -1;
  for (const candidate of ['\t', ';', ',']) {
    const counts = lines.map((l) => splitLine(l, candidate).length);
    if (Math.max(...counts) < 2) continue;
    const score = mostCommonCount(counts);
    if (score > bestScore) { bestScore = score; best = candidate; }
  }
  return best;
}

export function splitRows(text: string, delimiter: DelimiterOption = 'auto'): string[][] {
  const lines = text
    .replace(/\r\n?/g, '\n')
    .split('\n')
    .filter((l) => l.trim().length > 0);
  if (!lines.length) return [];

  const delim = delimiter === 'auto' ? detectDelimiter(lines) : delimiter;
  const rows = lines.map((l) => splitLine(l, delim));
  const width = Math.max(...rows.map((r) => r.length));
  return rows.map((r) => (r.length < width ? [...r, ...Array(width - r.length).fill('')] : r));
}

export function parseAmount(raw: string): number | null {
  if (raw == null) return null;
  let s = String(raw).trim();
  if (!s) return null;

  let negative = false;
  if (/^\(.*\)$/.test(s)) { negative = true; s = s.slice(1, -1); }
  s = s.replace(/[^0-9.,-]/g, '');
  if (s.startsWith('-')) { negative = true; s = s.slice(1); }
  if (s.endsWith('-')) { negative = true; s = s.slice(0, -1); }
  s = s.replace(/-/g, '');
  if (!s) return null;

  const lastComma = s.lastIndexOf(',');
  const lastDot = s.lastIndexOf('.');
  let normalized: string;

  if (lastComma !== -1 && lastDot !== -1) {
    normalized = lastComma > lastDot
      ? s.replace(/\./g, '').replace(',', '.')
      : s.replace(/,/g, '');
  } else if (lastComma !== -1) {
    const decimals = s.length - lastComma - 1;
    const isDecimal = decimals === 2 && (s.match(/,/g) ?? []).length === 1;
    normalized = isDecimal ? s.replace(',', '.') : s.replace(/,/g, '');
  } else if (lastDot !== -1) {
    const decimals = s.length - lastDot - 1;
    const dotCount = (s.match(/\./g) ?? []).length;
    normalized = dotCount > 1 || decimals === 3 ? s.replace(/\./g, '') : s;
  } else {
    normalized = s;
  }

  const n = Number(normalized);
  if (!Number.isFinite(n)) return null;
  return negative ? -n : n;
}

function toISODate(year: number, month: number, day: number): string | null {
  if (month < 1 || month > 12 || day < 1 || day > 31) return null;
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${year}-${pad(month)}-${pad(day)}`;
}

export function parseDate(raw: string): string | null {
  if (raw == null) return null;
  const s = String(raw).trim();
  if (!s) return null;

  let m = s.match(/^(\d{4})[-/](\d{1,2})[-/](\d{1,2})/);
  if (m) return toISODate(+m[1], +m[2], +m[3]);

  m = s.match(/^(\d{1,2})[/.-](\d{1,2})[/.-](\d{2,4})/);
  if (m) {
    const year = +m[3] < 100 ? 2000 + +m[3] : +m[3];
    return toISODate(year, +m[2], +m[1]);
  }
  return null;
}

export function guessRoles(rows: string[][]): ColumnRole[] {
  const width = rows[0]?.length ?? 0;
  const roles: ColumnRole[] = Array(width).fill('ignore');

  const scoreCol = (col: number, test: (v: string) => boolean) =>
    rows.reduce((n, r) => n + (test(r[col]) ? 1 : 0), 0);

  let dateCol = -1, dateScore = 0;
  for (let c = 0; c < width; c++) {
    const s = scoreCol(c, (v) => parseDate(v) !== null);
    if (s > dateScore) { dateScore = s; dateCol = c; }
  }
  if (dateCol >= 0 && dateScore > 0) roles[dateCol] = 'date';

  let amountCol = -1, amountScore = 0;
  for (let c = 0; c < width; c++) {
    if (c === dateCol) continue;
    const s = scoreCol(c, (v) => /\d/.test(v) && parseAmount(v) !== null);
    if (s > amountScore) { amountScore = s; amountCol = c; }
  }
  if (amountCol >= 0 && amountScore > 0) roles[amountCol] = 'amount';

  const descCol = roles.findIndex((r) => r === 'ignore');
  if (descCol >= 0) roles[descCol] = 'description';

  return roles;
}

export function looksLikeHeaderRow(rows: string[][], roles: ColumnRole[]): boolean {
  if (rows.length < 2) return false;
  const checkCols = roles
    .map((role, i) => ({ role, i }))
    .filter((c) => c.role === 'date' || c.role === 'amount');
  if (!checkCols.length) return false;

  const rowParses = (r: string[]) =>
    checkCols.some(({ role, i }) => (role === 'date' ? parseDate(r[i]) : parseAmount(r[i])) !== null);

  return !rowParses(rows[0]) && rows.slice(1).some(rowParses);
}

File: lib/parse_table.dart
import 'package:centavoo/models/transaction.dart' show kindExpense, kindRefund;

const colDate = 'date';
const colDescription = 'description';
const colAmount = 'amount';
const colIgnore = 'ignore';

const delimiterAuto = 'auto';
const delimiterComma = ',';
const delimiterSemicolon = ';';
const delimiterTab = '\t';

String deriveKind(double? amount) => (amount ?? 0) < 0 ? kindRefund : kindExpense;

bool deriveIsIof(String kind, String description) =>
    kind == kindRefund && description.toLowerCase().contains('iof');

List<String> _splitLine(String line, String delimiter) {
  final out = <String>[];
  var cur = StringBuffer();
  var inQuotes = false;
  for (var i = 0; i < line.length; i++) {
    final c = line[i];
    if (inQuotes) {
      if (c == '"') {
        if (i + 1 < line.length && line[i + 1] == '"') {
          cur.write('"');
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        cur.write(c);
      }
    } else if (c == '"') {
      inQuotes = true;
    } else if (c == delimiter) {
      out.add(cur.toString().trim());
      cur = StringBuffer();
    } else {
      cur.write(c);
    }
  }
  out.add(cur.toString().trim());
  return out;
}

int _mostCommonCount(List<int> counts) {
  final tally = <int, int>{};
  for (final n in counts) {
    tally[n] = (tally[n] ?? 0) + 1;
  }
  return tally.values.reduce((a, b) => a > b ? a : b);
}

String _detectDelimiter(List<String> lines) {
  var best = delimiterComma;
  var bestScore = -1;
  for (final candidate in [delimiterTab, delimiterSemicolon, delimiterComma]) {
    final counts = lines.map((l) => _splitLine(l, candidate).length).toList();
    if (counts.reduce((a, b) => a > b ? a : b) < 2) continue;
    final score = _mostCommonCount(counts);
    if (score > bestScore) {
      bestScore = score;
      best = candidate;
    }
  }
  return best;
}

List<List<String>> splitRows(String text, {String delimiter = delimiterAuto}) {
  final lines = text
      .replaceAll(RegExp(r'\r\n?'), '\n')
      .split('\n')
      .where((l) => l.trim().isNotEmpty)
      .toList();
  if (lines.isEmpty) return [];

  final delim = delimiter == delimiterAuto ? _detectDelimiter(lines) : delimiter;
  final rows = lines.map((l) => _splitLine(l, delim)).toList();
  final width = rows.map((r) => r.length).reduce((a, b) => a > b ? a : b);
  return rows.map((r) => r.length < width ? [...r, ...List.filled(width - r.length, '')] : r).toList();
}

double? parseAmount(String raw) {
  var s = raw.trim();
  if (s.isEmpty) return null;

  var negative = false;
  if (RegExp(r'^\(.*\)$').hasMatch(s)) {
    negative = true;
    s = s.substring(1, s.length - 1);
  }
  s = s.replaceAll(RegExp(r'[^0-9.,-]'), '');
  if (s.startsWith('-')) {
    negative = true;
    s = s.substring(1);
  }
  if (s.endsWith('-')) {
    negative = true;
    s = s.substring(0, s.length - 1);
  }
  s = s.replaceAll('-', '');
  if (s.isEmpty) return null;

  final lastComma = s.lastIndexOf(',');
  final lastDot = s.lastIndexOf('.');
  String normalized;

  if (lastComma != -1 && lastDot != -1) {
    normalized = lastComma > lastDot ? s.replaceAll('.', '').replaceAll(',', '.') : s.replaceAll(',', '');
  } else if (lastComma != -1) {
    final decimals = s.length - lastComma - 1;
    final commaCount = RegExp(',').allMatches(s).length;
    final isDecimal = decimals == 2 && commaCount == 1;
    normalized = isDecimal ? s.replaceAll(',', '.') : s.replaceAll(',', '');
  } else if (lastDot != -1) {
    final decimals = s.length - lastDot - 1;
    final dotCount = RegExp(r'\.').allMatches(s).length;
    normalized = dotCount > 1 || decimals == 3 ? s.replaceAll('.', '') : s;
  } else {
    normalized = s;
  }

  final n = double.tryParse(normalized);
  if (n == null || !n.isFinite) return null;
  return negative ? -n : n;
}

String? _toIsoDate(int year, int month, int day) {
  if (month < 1 || month > 12 || day < 1 || day > 31) return null;
  String pad(int n) => n.toString().padLeft(2, '0');
  return '$year-${pad(month)}-${pad(day)}';
}

String? parseDate(String raw) {
  final s = raw.trim();
  if (s.isEmpty) return null;

  var m = RegExp(r'^(\d{4})[-/](\d{1,2})[-/](\d{1,2})').firstMatch(s);
  if (m != null) {
    return _toIsoDate(int.parse(m.group(1)!), int.parse(m.group(2)!), int.parse(m.group(3)!));
  }

  m = RegExp(r'^(\d{1,2})[/.-](\d{1,2})[/.-](\d{2,4})').firstMatch(s);
  if (m != null) {
    final yearRaw = int.parse(m.group(3)!);
    final year = yearRaw < 100 ? 2000 + yearRaw : yearRaw;
    return _toIsoDate(year, int.parse(m.group(2)!), int.parse(m.group(1)!));
  }
  return null;
}

List<String> guessRoles(List<List<String>> rows) {
  final width = rows.isNotEmpty ? rows[0].length : 0;
  final roles = List<String>.filled(width, colIgnore);

  int scoreCol(int col, bool Function(String) test) =>
      rows.fold(0, (n, r) => n + (test(r[col]) ? 1 : 0));

  var dateCol = -1;
  var dateScore = 0;
  for (var c = 0; c < width; c++) {
    final s = scoreCol(c, (v) => parseDate(v) != null);
    if (s > dateScore) {
      dateScore = s;
      dateCol = c;
    }
  }
  if (dateCol >= 0 && dateScore > 0) roles[dateCol] = colDate;

  var amountCol = -1;
  var amountScore = 0;
  for (var c = 0; c < width; c++) {
    if (c == dateCol) continue;
    final s = scoreCol(c, (v) => RegExp(r'\d').hasMatch(v) && parseAmount(v) != null);
    if (s > amountScore) {
      amountScore = s;
      amountCol = c;
    }
  }
  if (amountCol >= 0 && amountScore > 0) roles[amountCol] = colAmount;

  final descCol = roles.indexOf(colIgnore);
  if (descCol >= 0) roles[descCol] = colDescription;

  return roles;
}

bool looksLikeHeaderRow(List<List<String>> rows, List<String> roles) {
  if (rows.length < 2) return false;
  final checkCols = [
    for (var i = 0; i < roles.length; i++)
      if (roles[i] == colDate || roles[i] == colAmount) (role: roles[i], i: i),
  ];
  if (checkCols.isEmpty) return false;

  bool rowParses(List<String> r) =>
      checkCols.any((c) => (c.role == colDate ? parseDate(r[c.i]) : parseAmount(r[c.i])) != null);

  return !rowParses(rows[0]) && rows.skip(1).any(rowParses);
}
